app/frontend/src/features/userCache/useUsers.ts

```diff
@@ -0,0 +1,96 @@
+import { Error } from "grpc-web";
+import { useQueries, useQueryClient } from "react-query";
+import { User } from "../../pb/api_pb";
+import { service } from "../../service";
+
+const userStaleTime = 10 * 60 * 1000;
+const username2IdStaleTime = 30 * 24 * 60 * 60 * 1000;
+
+export default function useUsers(
+  users: (number | string)[],
+  invalidate: boolean = false
+) {
+  const queryClient = useQueryClient();
+
+  //first get the ids of the usernames. This is doubling up, but this query
+  //doesn't become stale for a long time.
+  const usernames = users.filter((u): u is string => typeof u === "string");
+  const usernameQueries = useQueries<
+    { username: string; userId: number },
+    Error
+  >(
+    usernames.map((username) => ({
+      queryKey: ["username2Id", username],
+      queryFn: async () => {
+        const user = await service.user.getUser(username);
+        return { username: user.username, userId: user.userId };
+      },
+      staleTime: username2IdStaleTime,
+      cacheTime: username2IdStaleTime,
+    }))
+  );
+
+  //the query will only continue if all the ids of usernames are fetched
+  const isUserIdsFetched = usernameQueries.every((query) => query.data?.userId);
+
+  //filter only fetched ids from usernames, the query won't continue anyway
+  //if not all are fetched yet
+  const usernameIds = usernameQueries
+    .map((q) => q.data?.userId)
+    .filter((u): u is number => typeof u === "number");
+
+  const ids = users
+    .filter((u): u is number => typeof u === "number")
+    .concat(usernameIds);
+
+  if (invalidate && isUserIdsFetched) {
+    queryClient.invalidateQueries({
+      predicate: (query) =>
+        query.queryKey[0] === "user" &&
+        users.indexOf(query.queryKey[1] as number) > -1,
+    });
+  }
+
+  const queries = useQueries<User.AsObject, Error>(
+    ids.map((id) => ({
+      queryKey: ["user", id],
+      queryFn: () => service.user.getUser(id.toString()),
+      staleTime: userStaleTime,
+      enabled: isUserIdsFetched,
+    }))
+  );
+
+  const errors = [
+    usernameQueries
+      .map((query) => query.error?.message)
+      .filter((e): e is string => typeof e === "string"),
+    ...queries
+      .map((query) => query.error?.message)
+      .filter((e): e is string => typeof e === "string"),
+  ];
+
+  const isLoading =
+    queries.some((query) => query.isLoading) ||
+    usernameQueries.some((query) => query.isLoading);
+  const isError = !errors.length;
+
+  const usersById = new Map(queries.map((q) => [q.data?.userId, q.data]));
+  const usersByUsername = usernameQueries.map<
+    [string | undefined, User.AsObject | undefined]
+  >((q) => [
+    q.data?.username,
+    q.data ? usersById.get(q.data.userId) : undefined,
+  ]);
+
+  const allUsers = new Map<
+    number | string | undefined,
+    User.AsObject | undefined
+  >([...Array.from(usersById), ...usersByUsername]);
+
+  return {
+    isLoading,
+    isError,
+    errors,
+    users: allUsers,
+  };
```

Comment:
@darrenvong what do you think of this approach?

It's works for the general case and mixing usernames and ids. But after writing it, I think it's too bloody verbose and annoying.

Basically, it:

- separates usernames and userIds based on type
- runs a query on the usernames which returns a username and userId object. This query doesn't go stale (since usernames can't change) for a very long time.
- joins those ids with the ids passed as a parameter
- runs the main query to fetch on id
- returns a map of userIds and usernames -> user objects.

As far as I remember, the only time we get the user by username is for the user page, right? In that case, maybe I'm better off doing the same approach (having a separate username -> userId query, the running then normal id query) but moving it out into a singular `useUserByUsername(username: string)` until we need to be able to get multiple by username (which may be never).

Reply:
Yeah it's a bit annoying/awkward that the users page are by username.

When you say move it to a separate `useUserByUsername` hook, do you mean it will still try to store users under the `["user", userId]` key?

Reply:
Yes exactly. I'll give it a try, I think it will be much nearer.